include/sdf/ParserConfig.hh

```diff
@@ -0,0 +1,130 @@
+///
+/// For backward compatibility, the functions sdf::setFindCallback() and
+/// sdf::addURIPath() update a singleton ParserConfig object, which can be
+/// retrieved by ParserConfig::GlobalConfig().
+///
+/// The functions sdf::readFile(), sdf::readString(), and \ref Root::Load
+/// "sdf::Root::Load()" have overloads that take a ParserConfig object. If
+/// the ParserConfig object is omitted, these functions will use the singleton
+/// ParserConfig object.
+///
+/// Example:
+/// To set an additional URI scheme search directory without affecting the
+/// global config,
+///
+/// \code{.cpp}
+///   // Copy the global config
+///   sdf::ParserConfig config = ParserConfig::GlobalConfig();
+///
+///   // Add the new scheme to the config
+///   config.AddURIPath("newScheme://", "path/to/directory");
+///
+///   // Use the new config when loading a new SDFormat file
+///   sdf::Root root;
+///   root.Load("path/to/file.sdf", config);
+/// \endcode
+class SDFORMAT_VISIBLE ParserConfig
+{
+  /// type alias for the map from URI scheme to search directories
+  public: using SchemeToPathMap =
+          std::map<std::string, std::vector<std::string> >;
+
+  /// \brief Default constructor
+  public: ParserConfig();
+
+  /// \brief Copy constructor
+  /// \param[in] _config ParserConfig to copy.
+  public: ParserConfig(const ParserConfig &_config);
+
+  /// \brief Move constructor
+  /// \param[in] _config ParserConfig to move.
+  public: ParserConfig(ParserConfig &&_config) noexcept;
+
+  /// \brief Move assignment operator.
+  /// \param[in] _config ParserConfig to move.
+  /// \return Reference to this.
+  public: ParserConfig &operator=(ParserConfig &&_config) noexcept;
+
+  /// \brief Copy assignment operator.
+  /// \param[in] _config ParserConfig to copy.
+  /// \return Reference to this.
+  public: ParserConfig &operator=(const ParserConfig &_config);
+
+  /// \brief Destructor
+  public: ~ParserConfig();
+
+  /// Mutable access to a singleton ParserConfig that serves as the global
+  /// ParserConfig object for all parsing operations
+  /// \return A mutable reference to the singleton ParserConfig object
+  public: static ParserConfig &GlobalConfig();
+
+  /// \brief Get the find file callback function
+  /// \return Immutable reference to the find file callback function
+  public: const std::function<std::string(const std::string &)> &
+          FindFileCallback() const;
+
+  /// \brief Set the callback to use when libsdformat can't find a file.
+  /// The callback should return a complete path to the requested file, or
+  /// and empty string if the file was not found in the callback.
+  /// \param[in] _cb The callback function.
+  public: void SetFindCallback(
+              std::function<std::string(const std::string &)> _cb);
+
+  /// \brief Get the URI scheme to search directories map
+  /// \return Immutable reference to the URI scheme to search directories map
+  public: const SchemeToPathMap &URIPathMap() const;
```

Comment:
nit It smells weird to be able to look at all paths, add a new path, but not remove a path. (assymmetry)

Is it hard to add that here?

Or perhaps the right workflow is to just reassign? e.g. just clear it out.

So if I wanted to remove or reorder a path, I would just do something like `Parser::GLobalConfig() = my_new_config`?

Reply:
I was trying to make this PR replicate existing functionality without adding any new features other than being able to specify a non-global config. It would be easy to add a function to remove a URI scheme, but adding a path to an existing URI scheme or reordering paths seem best suited for a future PR, IMO.

Reply:
Aye, makes sense to me.